### ISSUE TITLE
SNOW-1524204: Session Variables passed as connection params are ignored

### ISSUE THREAD

**Issue by OrCh3n** [snowflakedb/gosnowflake#1173]
1. What version of GO driver are you using?
v1.7.2
   
2. What operating system and processor architecture are you using?
darwin/arm64
   
3. What version of GO are you using?
1.21.8

4.Server version:* E.g. 1.90.1
You may get the server version by running a query:
8.24.1

5. What did you do?
```
		sfCfg = &gosnowflake.Config{
			<account, user etc.>
			Params: map[string]*string{"$customer_id": lo.ToPtr("test")}
		}
		url, _ = gosnowflake.DSN(sfCfg)
		// open the db using the url
                // execute SHOW VARIABLES -> nothing shows
```
6. What did you expect to see?
as per the docs https://docs.snowflake.com/en/sql-reference/session-variables
there should be a session variable named CUSTOMER_ID.

The issue is the $ is url encoded, so when parsing the config back from the url it is left url encoded, while it needs to be a $ sign.
ParseDSN should decode the parameter keys IMO.

**Comment by sfc-gh-dszmolka** [comment]
hey, thanks for raising this ! so the `Params` field is supposed to contain only connection and session _parameters_ , hence the name. Please see [this doc](https://pkg.go.dev/github.com/snowflakedb/gosnowflake#hdr-Connection_Parameters) for reference.

if you're trying to use a session _variable_ , which can be indeed be anything, can you please try specifying them per the documentation you linked; using the `SET` statement after establishing the connection? 
For example, `SET customer_id='test'`

**Comment by sfc-gh-dszmolka** [comment]
as you mentioned, the session variable indeed comes urlencoded
```
        "SESSION_PARAMETERS": {
            "%24CUSTOMER_ID": "test",
            "CLIENT_VALIDATE_DEFAULT_PARAMETERS": true
        },
```

and thus never actually recognized as a session variable, by the server. As a workaround, please consider using the method mentioned previously while we work on this enhancement for the driver to be on par with the JDBC driver regarding this aspect.

**Comment by sfc-gh-dszmolka** [comment]
hopefully https://github.com/snowflakedb/gosnowflake/pull/1177 can address this. Thank you again for pointing this gap out !

**Comment by OrCh3n** [comment]
Thanks for the quick response!

**Comment by sfc-gh-dszmolka** [comment]
fix PR is now merged and will be part of the next upcoming release

**Comment by sfc-gh-dszmolka** [comment]
released with gosnowflake v1.11.0 in July 2024
